### ISSUE TITLE
CentOS 6.9 - Failed to update secret file

### ISSUE THREAD

**Issue by fmgdias** [google/google-authenticator-libpam#93]
I'm not able to activate google-authenticator on my CentOS server, can you guys help me?

```

[root@xxx log]# cat /etc/redhat-release 
CentOS release 6.9 (Final)

[root@xxx log]# tail -f secure
Mar  8 14:00:56 xxx sshd[26942]: Connection closed by 201.6.1.20
Mar  8 14:01:02 xxx sshd(pam_google_authenticator)[26946]: Accepted google_authenticator for fmgodias
Mar  8 14:01:02 xxx sshd(pam_google_authenticator)[26946]: Failed to delete tempfile "/home/fmgodias/.google_authenticator~O40FNf": No such file or directory
Mar  8 14:01:02 xxx sshd(pam_google_authenticator)[26946]: Failed to update secret file "/home/fmgodias/.google_authenticator": Permission denied

[root@xxx tmp]# cat /etc/pam.d/sshd
#%PAM-1.0
auth       required     pam_google_authenticator.so nullok
auth	   required	pam_sepermit.so
auth       include      password-auth
account    required     pam_nologin.so
account    include      password-auth
password   include      password-auth
# pam_selinux.so close should be the first session rule
session    required     pam_selinux.so close
session    required     pam_loginuid.so
# pam_selinux.so open should only be followed by sessions to be executed in the user context
session    required     pam_selinux.so open env_params
session    required     pam_namespace.so
session    optional     pam_keyinit.so force revoke
session    include      password-auth

[root@xxx tmp]# getenforce
Enforcing

```

Install How-To
```
### Compile Google Authenticator
yum remove google-authenticator -y;
yum install qrencode pam-devel make gcc wget autoreconf unzip autoconf automake libtool -y;
mkdir -p /opt/google-authenticator/ ;
cd /opt/google-authenticator/ ;
wget https://github.com/google/google-authenticator-libpam/archive/master.zip -O google-authenticator-libpam.zip ;
unzip google-authenticator-libpam.zip;
cd /opt/google-authenticator/google-authenticator* ;
./bootstrap.sh ;
./configure ;
make && make install ;

# Install Google Authenticator
unlink /usr/bin/google-authenticator ;
unlink /lib64/security/pam_google_authenticator.so ;
unlink /lib64/security/pam_google_authenticator.la ;
ln -s /usr/local/bin/google-authenticator /usr/bin/google-authenticator ;
ln -s /usr/local/lib/security/pam_google_authenticator.so /lib64/security/pam_google_authenticator.so ;
ln -s /usr/local/lib/security/pam_google_authenticator.la /lib64/security/pam_google_authenticator.la ;

# Config SSH Daemon
sed -i '/#%PAM/a auth\ \ \ \ \ \ \ required\ \ \ \ \ pam_google_authenticator.so nullok' /etc/pam.d/sshd ;
sed -i 's/#ChallengeResponseAuthentication\ yes/ChallengeResponseAuthentication\ yes/g' /etc/ssh/sshd_config ;
sed -i 's/ChallengeResponseAuthentication\ no/#ChallengeResponseAuthentication\ no/g' /etc/ssh/sshd_config ;
sed -i 's/#UseDNS\ yes/UseDNS\ no/g' /etc/ssh/sshd_config ;
sed -i 's/#PermitRootLogin\ yes/PermitRootLogin\ no/g' /etc/ssh/sshd_config ;
service sshd restart ;

```

**Comment by ThomasHabets** [comment]
Who owns `/home/fmgodias/.google_authenticator` and `/home/fmgodias`, and what permissions do they have?

I'm not very familiar with SELinux, so if it's not an ownership thing then it's probably SELinux blocking it somehow.

**Comment by fmgdias** [comment]
it's because SELinux is active, in the past I've been able to create a release rule using the URL hints below, but it does not work anymore.
If I disable SELinux it works, but I would like a solution with SELinux active.

http://blog.boa.nu/2012/11/two-factor-ssh-login-google-authenticator-and-selinux.html

**Comment by drussell393** [comment]
@fmgdias,

I just setup Google Auth with SELinux enabled on my CentOS 7.x box. I used parts of that link, but found that there was an update ages ago in SELinux that attempted to automate the Google Authenticator process, which made name type conflicts come up when I was trying to make my own policy/type.

That being said, I succeeded. However, there's only two ways (as far as I can tell) to succeed in this because of those name type conflicts. Neither are as pretty as I'd like, but it is what it is.


There's two things to keep in mind when you're making your policy:

1. SELinux doesn't give `sshd_t` access to `user_home_dir_t` by default, and you'll need to give `sshd_t` write & read access to this (as well as setattr, getattr, etc) type in order for the `.google_authenticator` file to be directly in the user's home directory

Theoretically, you could move this file into its own directory, and then give that directory the context type `ssh_home_t` to get around that.

2. SELinux doesn't, by default, give the right context type to a newly created file without passing `filetrans_pattern` in the type configuration

**Option 1 - Renaming ".google_authenticator" to something else in the home dir**

You could either move `.google_authenticator` into a new directory, thereby getting rid of the rule that SELinux had by default (`/home/[^/]+/\.google_authenticator` takes on `auth_home_t` by default), or you could change `.google-authenticator` (the file name) to something else.

From there, you'd follow all of the steps in that URL to get the custom type setup, but keep in mind that you'll also need to pass the following "allow" unless you're working in a `ssh_home_t` context directory:

`allow sshd_t user_home_dir_t : file { read write rename create unlink rw_file_perms getattr setattr };`

**Option 2 - Give all files changed in `user_home_dir_t` by `sshd_t` the custom context you make**

This is easier if you're not relying on authorized_keys and other keys being inside of the user's home directory. My keys are all outside of the home directory (and authorized_keys are on a separate system partition), so this works for me.

Basically, instead of passing:

`filetrans_pattern(sshd_t, user_home_dir_t, CUSTOM_TYPE_NAME_T, file, '.google_authenticator')`

You would pass:

`filetrans_pattern(sshd_t, user_home_dir_t, CUSTOM_TYPE_NAME_T, file)`

This would also get around the fact that libpam-google-authenticator automatically attempts to create a temporary file (usually `.google-authenticator~` with a random code at the end). Because you can't pass a regex pattern into `filetrans_pattern`, this would avoid any AVC errors caused by that create.

-- Dave.
